@here/harp-mapview/lib/Utils.ts

```diff
@@ -242,53 +243,190 @@ export namespace MapViewUtils {
         deltaTilt: number,
         maxTiltAngle: number
     ) {
-        const camera = mapView.camera;
-        const projection = mapView.projection;
-
         const rotationTargetWorld = MapViewUtils.rayCastWorldCoordinates(mapView, offsetX, offsetY);
         if (rotationTargetWorld === null) {
             return;
         }
 
-        const headingAxis = projection.surfaceNormal(rotationTargetWorld, cache.vector3[1]);
-        const headingQuat = cache.quaternions[1].setFromAxisAngle(headingAxis, -deltaAzimuth);
+        applyAzimuthAroundTarget(mapView, rotationTargetWorld, -deltaAzimuth);
+
+        const clampedDeltaTilt = computeClampedDeltaTilt(
+            mapView,
+            offsetX,
+            offsetY,
+            deltaTilt,
+            maxTiltAngle,
+            rotationTargetWorld
+        );
+
+        applyTiltAroundTarget(mapView, rotationTargetWorld, clampedDeltaTilt);
+    }
+
+    /**
+     * @hidden
+     * @internal
+     *
+     * Applies the given Azimith to the camera around the supplied target.
+     */
+    function applyAzimuthAroundTarget(
+        mapView: MapView,
+        rotationTargetWorld: THREE.Vector3,
+        deltaAzimuth: number
+    ) {
+        const camera = mapView.camera;
+        const projection = mapView.projection;
+
+        const headingAxis = projection.surfaceNormal(rotationTargetWorld, cache.vector3[0]);
+        const headingQuat = cache.quaternions[0].setFromAxisAngle(headingAxis, deltaAzimuth);
         camera.quaternion.premultiply(headingQuat);
         camera.position.sub(rotationTargetWorld);
         camera.position.applyQuaternion(headingQuat);
         camera.position.add(rotationTargetWorld);
+    }
+
+    /**
+     * @hidden
+     * @internal
+     *
+     * Clamps the supplied `deltaTilt` to the `maxTiltAngle` supplied. Note, when a non-zero offset
+     * is applied, we apply
+     */
+    function computeClampedDeltaTilt(
+        mapView: MapView,
+        offsetX: number,
+        offsetY: number,
+        deltaTilt: number,
+        maxTiltAngle: number,
+        rotationTargetWorld: THREE.Vector3
+    ): number {
+        const camera = mapView.camera;
+        const projection = mapView.projection;
 
         const mapTargetWorld =
             offsetX === 0 && offsetY === 0
                 ? rotationTargetWorld
                 : MapViewUtils.rayCastWorldCoordinates(mapView, 0, 0);
         if (mapTargetWorld === null) {
-            return;
+            return 0;
         }
```

Comment:
Before we were returning from "orbitAroundScreenPoint" if the map target is not on the ground. Now we return 0 tilt and go on an apply it. We could return immediately as before.
Maybe this piece of code belongs to "orbitAroundScreenPoint" instead, that way you can return from the function directly without forcing this one to return some undefined tilt value.

Reply:
I will move.